network/local.go

```diff
@@ -0,0 +1,487 @@
+func (lm *LocalManager) connect(local, remote Address) (*LocalConn, error) {
+	lm.Lock()
+	defer lm.Unlock()
+
+	fn, ok := lm.listening[remote]
+	if !ok {
+		return nil, fmt.Errorf("%s can't connect to %s: it's not listening", local, remote)
+	}
+
+	outEndpoint := endpoint{local, lm.counter}
+	lm.counter++
+
+	incEndpoint := endpoint{remote, lm.counter}
+	lm.counter++
+
+	outgoing := newLocalConn(lm, outEndpoint, incEndpoint)
+	incoming := newLocalConn(lm, incEndpoint, outEndpoint)
+
+	// outgoing knows how to store packets into the incoming's queue.
+	lm.queues[outEndpoint] = outgoing.connQueue
+	// incoming knows how to store packets into the outgoing's queue.
+	lm.queues[incEndpoint] = incoming.connQueue
+
+	go fn(incoming)
+	return outgoing, nil
+}
+
+// send gets the connection denoted by this endpoint and calls queueMsg
+// with the packet as argument to it.
+// It returns ErrClosed if it does not find the connection.
+func (lm *LocalManager) send(e endpoint, buff Packet) error {
+	lm.Lock()
+	defer lm.Unlock()
+	q, ok := lm.queues[e]
+	if !ok {
+		return ErrClosed
+	}
+
+	q.push(buff)
+	return nil
+}
+
+// close gets the connection denoted by this endpoint and closes it if
+// it is present.
+func (lm *LocalManager) close(conn *LocalConn) {
+	lm.Lock()
+	defer lm.Unlock()
+	// delete this conn
+	delete(lm.queues, conn.local)
+	// and delete the remote one + close it
+	remote, ok := lm.queues[conn.remote]
+	if !ok {
+		return
+	}
+	delete(lm.queues, conn.remote)
+	remote.close()
+}
+
+// len returns how many local connections are open.
+func (lm *LocalManager) len() int {
+	lm.Lock()
+	defer lm.Unlock()
+	return len(lm.queues)
+}
+
+// LocalConn is a connection that sends and receives messages to other
+// connections locally.
+type LocalConn struct {
+	local  endpoint
+	remote endpoint
+
+	// connQueue is accessible from the LocalManager (i.e. is
+	// shared). We can't directly share LocalConn because go test
+	// -race detects it as data race (while it's *protected*).
+	*connQueue
+
+	// counter to keep track of how many bytes read / written this connection
+	// has seen.
+	counterSafe
+	// the localManager responsible for that connection.
+	manager *LocalManager
+}
+
+// newLocalConn initializes the fields of a LocalConn but does'nt
+// connect. It should not be used from the outside, most user want
+// to use NewLocalConn.
+func newLocalConn(lm *LocalManager, local, remote endpoint) *LocalConn {
+	return &LocalConn{
+		remote:    remote,
+		local:     local,
+		connQueue: newConnQueue(),
+		manager:   lm,
+	}
+}
+
+// NewLocalConn returns a new channel connection from local to remote.
+// It mimics the behavior of NewTCPConn and tries to connect right away.
+// It uses the default local manager.
+func NewLocalConn(local, remote Address) (*LocalConn, error) {
+	return NewLocalConnWithManager(defaultLocalManager, local, remote)
+}
+
+// NewLocalConnWithManager is similar to NewLocalConn but takes a specific
+// LocalManager.
```

Comment:
Not really. It starts the listening go routine from the other endpoint, so *like tcp*: if you have a tcp listener, and then tries to connect to it, a new go routine will be launched. Not from the connect-er perspective but from the connect-ee.

I could put a comment here but I'm not sure it would be less confusing than anything else... (as an example, is my comment above clear ?)